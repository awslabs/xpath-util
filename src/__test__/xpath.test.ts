import { JSDOM } from "jsdom";

import {
    getXpath
} from "../xpath";

export const sampleHTML = 
`
<div id="another-better-id">
    <div 
        id="some-unique-id"
        data-analytics="otherParent"
        data-analytics-type="serviceSubSection"
    >
        <div 
            class="ASKLJHASDLKJH" 
            style="background: #eee; padding: 10px"
            data-analytics="introParagraph"
        >
            <p>
                This is a sibling paragraph
            </p>
            <p>
            This paragraph
            contains      a lot of spaces
            in the source     code,
            but the    browser 
            ignores it.
            </p>

        </div>
    </div>
</div>
`;

test("Can get element info from block of HTML", () => {
    const inputElement = document.createElement("div");
    inputElement.innerHTML = sampleHTML;
    document.body.appendChild(inputElement);

    // We'll grab the paragraph by diving into the element twice
    // using the "firstChild" attribute
    const firstParagraph = (document.querySelector("#another-better-id p") as HTMLElement);
    const secondParagraph = (document.querySelector("#another-better-id p:last-child") as HTMLElement);

    const firstParagraphElementInfo = getXpath(firstParagraph)

    // A catchall snapshot
    expect(firstParagraphElementInfo).toMatchSnapshot();

    // More precise tests around the returned element data
    expect(firstParagraphElementInfo.elementData["tag"]).toBe("introParagraph")
    expect(firstParagraphElementInfo.elementData["serviceSubSection"]).toBe("otherParent")

    const secondParagraphElementInfo = getXpath(secondParagraph)
    expect(secondParagraphElementInfo).toMatchSnapshot();
});

test("Test what happens when passing an empty element", () => {
    const inputElement = document.createElement("div");

    const elementInfo = getXpath(inputElement);

    expect(elementInfo.customPath).toBe("");
    expect(elementInfo.elementData).toStrictEqual({});
})