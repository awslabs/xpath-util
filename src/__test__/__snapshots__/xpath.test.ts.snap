// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Can get element info from block of HTML 1`] = `
Object {
  "customPath": "/serviceSubSection(otherParent)/tag(introParagraph)",
  "elementData": Object {
    "serviceSubSection": "otherParent",
    "tag": "introParagraph",
  },
  "xpath": "/html/body/div/div[@id=\\"another-better-id\\"]/div[@id=\\"some-unique-id\\"]/div[@class=\\"ASKLJHASDLKJH\\"]/p",
}
`;

exports[`Can get element info from block of HTML 2`] = `
Object {
  "customPath": "/serviceSubSection(otherParent)/tag(introParagraph)",
  "elementData": Object {
    "serviceSubSection": "otherParent",
    "tag": "introParagraph",
  },
  "xpath": "/html/body/div/div[@id=\\"another-better-id\\"]/div[@id=\\"some-unique-id\\"]/div[@class=\\"ASKLJHASDLKJH\\"]/p[1]",
}
`;
